plugin/evm/epochs_manager_test.go

```diff
@@ -0,0 +1,75 @@
+package evm
+
+import (
+	"fmt"
+	"math/big"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type testEpochs struct {
+	DetailsFunc func(epoch uint64) (EpochDetails, error)
+}
+
+func (t *testEpochs) Details(epoch uint64) (EpochDetails, error) {
+	return t.DetailsFunc(epoch)
+}
+
+func TestByTimestamp(t *testing.T) {
+	callsCount := 0
+	testEpochsMock := &testEpochs{
+		DetailsFunc: func(epoch uint64) (EpochDetails, error) {
+			callsCount++
+			epochDetails1 := EpochDetails{
+				StartTime: big.NewInt(time.Date(2022, time.January, 1, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+				EndTime:   big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+			}
+			epochDetails2 := EpochDetails{
+				StartTime: big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+				EndTime:   big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+			}
+			epochDetails3 := EpochDetails{
+				StartTime: big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+				EndTime:   big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+			}
+			epochDetails4 := EpochDetails{
+				StartTime: big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+				EndTime:   big.NewInt(time.Date(2022, time.December, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
+			}
+			if epoch < 0 {
+				return EpochDetails{}, fmt.Errorf("Invalid epoch")
+			}
+			switch epoch {
+			case 0:
+				return epochDetails1, nil
+			case 1:
+				return epochDetails2, nil
+			case 2:
+				return epochDetails3, nil
+			default:
+				return epochDetails4, nil
```

Comment:
Instead of doing this and making a mock with logic to handle multiple tests, please just overwrite the function with one that returns exactly what the test needs in each subtest.

Reply:
`DetailsFunc()` has to be called multiple times so that the `epoch` settles at the right range. I am not sure how returning one single thing will achieve this? Maybe if you can give an example it will be clearer.

Reply:
There is only one subtest where we need to call it multiple times, and only in this specific subtest we'll want to have something like

```go
testDetails := []EpochDetails{
			{
				StartTime: big.NewInt(time.Date(2022, time.January, 1, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
				EndTime:   big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
			},
			{
				StartTime: big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
				EndTime:   big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
			},
			{
				StartTime: big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
				EndTime:   big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
			},
			{
				StartTime: big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
				EndTime:   big.NewInt(time.Date(2022, time.December, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
			}
}

t.Run("finds correct epoch details among multiple calls", func(t *testing.T) {
	var called int
	mock := &testEpochs{
		DetailsFunc: func(epoch uint64) (EpochDetails, error) {
			called++
			return testDetails[i], nil
		},
	}

	// ...
}
```

Do you see what I mean? :)

Reply:
I understand better now but still not fully clear as I don't know what `i` is in your above suggestion. 
I am almost done for today, so if you have time later, feel free to push your suggestions. Thanks :)

Reply:
My bad, I edited the snippet, I meant for it to be `epoch`, and to use an epoch that matches the index in the slice.